fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java

```diff
@@ -60,97 +116,111 @@ public HDFSBackendSelector(ScanNode scanNode, List<TScanRangeLocations> location
         this.computeNodes = computeNodes;
         this.forceScheduleLocal = forceScheduleLocal;
         this.addressToBackendId = addressToBackendId;
-        this.scanRangesBytes =
-                locations.stream().map(x -> x.scan_range.hdfs_scan_range.length).collect(Collectors.toList());
+        this.hdfsScanRangeHasher = new HdfsScanRangeHasher();
+    }
+
+    private ComputeNode selectLeastScanBytesComputeNode(Collection<ComputeNode> backends) {
+        if (backends == null || backends.isEmpty()) {
+            return null;
+        }
+
+        ComputeNode node = null;
+        long minAssignedScanRanges = Long.MAX_VALUE;
+        for (ComputeNode backend : backends) {
+            long assignedScanRanges = assignedScansPerComputeNode.get(backend);
+            if (assignedScanRanges < minAssignedScanRanges) {
```

Comment:
Do we need to adjust assignment for balancing scan ranges in every single query.
First, the hash mechanism should assure the distribution is even.
Second, I'm afraid the adjustment will cause some scan range move from one node to another, causing cache miss.

Reply:
> Do we need to adjust assignment for balancing scan ranges in every single query. 
Yes, we do

> First, the hash mechanism should assure the distribution is even. 
In theory yes,  but we also have to consider bad cases. If distirbution is not even, then there are some hotspots, we don't want to see that. 

> Second, I'm afraid the adjustment will cause some scan range move from one node to another, causing cache miss.

We select at most 3 different nodes, so cache data will be replicated at most 3 times.  Cache miss is not the only factor we should consider, we should also consider parallelism.  There is a balance between them.

Reply:
Maybe we could try to do that after all scan range assigned. It's hard to know whether it's uniformly distributed before the assignment finish.

Reply:
I don't get it how it works, could you elaborate a little?

Let's say we have 4 scan ranges (S1,S2,S3, S4) and 4 hosts(H1 - H4).  For S1, we get H1-H3 as candidates, which host do we choose?

Reply:
I almost get what you mean. I'll write another version of it and you can review.